File: app/src/main/java/com/example/earthwallet/wallet/passport/PassportInputs.kt
package network.erth.wallet.wallet.passport

import org.bouncycastle.asn1.ASN1Integer
import org.bouncycastle.asn1.ASN1ObjectIdentifier
import org.bouncycastle.asn1.ASN1Sequence
import org.bouncycastle.asn1.cms.ContentInfo
import org.bouncycastle.cert.X509CertificateHolder
import org.bouncycastle.cms.CMSSignedData
import java.math.BigInteger
import java.security.MessageDigest

/**
 * PassportInputs — builds the inputs for the lean_poa Noir circuit
 * (earth-network-mobile/circuits/lean_poa) from a scanned passport's EF.DG1 and
 * EF.SOD, using BouncyCastle (already bundled) for the ASN.1/CMS work.
 *
 * Every input is derived from the passport itself. The circuit computes both the
 * nullifier (Poseidon2 over name+DOB) and the DSC commitment in-circuit and
 * returns them, so there is NO Poseidon2 needed in Kotlin and no registry
 * round-trip before proving — the Document Signer certificate simply travels with
 * MsgRegister for the chain to verify against its CSCA trust store.
 *
 * Circuit input contract (see circuits/lean_poa/SECURITY.md):
 *   dg1[95] + dg1_len, e_content[200] + e_content_len + dg1_hash_offset,
 *   signed_attrs[200] + signed_attrs_len + econtent_hash_offset,
 *   dsc_pubkey_x[32], dsc_pubkey_y[32], sod_signature[64] (low-s), current_date.
 */
object PassportInputs {

    /** An ECDSA DSC curve: its register-circuit id, coordinate byte length, and
     *  group order n (for low-s signature normalization). */
    private class EcCurve(val algorithm: String, val coordLen: Int, val order: BigInteger)

    private fun order(hex: String) = BigInteger(hex, 16)

    // Supported EC DSC curves, keyed by named-curve OID.
    private val EC_CURVES = mapOf(
        "1.2.840.10045.3.1.7" to EcCurve("lean_poa", 32, order("FFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551")), // P-256
        "1.3.132.0.34" to EcCurve("lean_poa_p384", 48, order("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFC7634D81F4372DDF581A0DB248B0A77AECEC196ACCC52973")), // P-384
        "1.3.36.3.3.2.8.1.1.7" to EcCurve("lean_poa_brainpool256", 32, order("A9FB57DBA1EEA9BC3E660A909D838D718C397AA3B561A6F7901E0E82974856A7")),
        "1.3.36.3.3.2.8.1.1.11" to EcCurve("lean_poa_brainpool384", 48, order("8CB91E82A3386D280F5D6F7E50E641DF152F7109ED5456B31F166E6CAC0425A7CF3AB6AF6B7FC3103B883202E9046565")),
        "1.3.36.3.3.2.8.1.1.13" to EcCurve("lean_poa_brainpool512", 64, order("AADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA70330870553E5C414CA92619418661197FAC10471DB1D381085DDADDB58796829CA90069")),
    )

    private const val DG1_MAX = 95
    private const val ECONTENT_MAX = 200
    private const val SIGNED_ATTRS_MAX = 200

    /** A DSC extracted from a passport's SOD: its cert DER + canonical public key. */
    class ScannedDsc(val certificateDer: ByteArray, val pubkey: ByteArray)

    /**
     * Extracts the Document Signer Certificate from the passport's EF.SOD and its
     * canonical public key (the x/pki dedup/lookup key and the Merkle-leaf preimage,
     * matching the chain: ECDSA -> x‖y, RSA -> modulus big-endian).
     */
    fun scannedDsc(sodBytes: ByteArray): ScannedDsc {
        val signedData = CMSSignedData(ContentInfo.getInstance(stripSodTag(sodBytes)))
        @Suppress("UNCHECKED_CAST")
        val dsc = (signedData.certificates.getMatches(null) as Collection<X509CertificateHolder>).first()
        val spki = dsc.subjectPublicKeyInfo
        val pubkey = if (spki.algorithm.algorithm.id == "1.2.840.10045.2.1") { // ecPublicKey
            val point = spki.publicKeyData.bytes // 0x04 || X || Y (uncompressed)
            val coordLen = (point.size - 1) / 2
            point.copyOfRange(1, 1 + 2 * coordLen)
        } else { // rsaEncryption: canonical = modulus big-endian
            org.bouncycastle.asn1.pkcs.RSAPublicKey.getInstance(spki.parsePublicKey()).modulus.toByteArrayUnsigned()
        }
        return ScannedDsc(dsc.encoded, pubkey)
    }

    /**
     * Builds the lean_poa circuit inputs as the flat map noir_android expects.
     *
     * @param dg1 raw EF.DG1 bytes (as read from the chip).
     * @param sodBytes raw EF.SOD bytes.
     * @param currentDateYymmdd today as a YYMMDD integer (e.g. 260813). The chain
     *   pins this to block time, so it must be ~now.
     * @param registry the DSC's registry inclusion proof.
     */
    /** The circuit inputs plus the register-circuit id to prove/verify against. */
    class Inputs(val algorithm: String, val map: Map<String, Any>)

    /**
     * Builds the register-circuit inputs for the passport's DSC algorithm and
     * selects the matching circuit: `lean_poa` (ECDSA-P256), `lean_poa_rsa2048`,
     * or `lean_poa_rsa4096`. The hash-binding, registry, and expiry inputs are the
     * same across variants (poa_core); only the DSC signature/key fields differ.
     */
    fun buildInputs(
        dg1: ByteArray,
        sodBytes: ByteArray,
        currentDateYymmdd: Int,
    ): Inputs {
        require(dg1.size <= DG1_MAX) { "DG1 is ${dg1.size} bytes, exceeds circuit max $DG1_MAX" }

        val signedData = CMSSignedData(ContentInfo.getInstance(stripSodTag(sodBytes)))
        val signer = signedData.signerInfos.signers.first()

        // eContent = the LDS security object (holds the DG hashes); signed_attrs =
        // the SignerInfo signed-attribute SET in its to-be-signed DER form.
        val eContent = signedData.signedContent.content as ByteArray
        val signedAttrs = signer.encodedSignedAttributes
        require(eContent.size <= ECONTENT_MAX) {
            "eContent is ${eContent.size} bytes, exceeds circuit max $ECONTENT_MAX (bump ECONTENT_MAX + circuit)"
        }
        require(signedAttrs.size <= SIGNED_ATTRS_MAX) {
            "signed attributes are ${signedAttrs.size} bytes, exceeds circuit max $SIGNED_ATTRS_MAX"
        }

        // Hash bindings: sha256(dg1) sits in eContent; sha256(eContent) sits in
        // the signed attributes (the messageDigest attribute).
        val dg1HashOffset = indexOfSubarray(eContent, sha256(dg1))
        require(dg1HashOffset >= 0) { "sha256(DG1) not found in eContent" }
        val econtentHashOffset = indexOfSubarray(signedAttrs, sha256(eContent))
        require(econtentHashOffset >= 0) { "sha256(eContent) not found in signed attributes" }

        // Shared hash-binding inputs (poa_core).
        val map = LinkedHashMap<String, Any>()
        map["dg1"] = byteArrayInput(pad(dg1, DG1_MAX))
        map["dg1_len"] = scalarInput(dg1.size)
        map["e_content"] = byteArrayInput(pad(eContent, ECONTENT_MAX))
        map["e_content_len"] = scalarInput(eContent.size)
        map["dg1_hash_offset"] = scalarInput(dg1HashOffset)
        map["signed_attrs"] = byteArrayInput(pad(signedAttrs, SIGNED_ATTRS_MAX))
        map["signed_attrs_len"] = scalarInput(signedAttrs.size)
        map["econtent_hash_offset"] = scalarInput(econtentHashOffset)

        // DSC-algorithm-specific signature/key fields + circuit selection.
        @Suppress("UNCHECKED_CAST")
        val dsc = (signedData.certificates.getMatches(null) as Collection<X509CertificateHolder>).first()
        val spki = dsc.subjectPublicKeyInfo
        val algorithm = when (spki.algorithm.algorithm.id) {
            "1.2.840.10045.2.1" -> { // ecPublicKey — select the curve variant
                val curveOid = ASN1ObjectIdentifier.getInstance(spki.algorithm.parameters).id
                val curve = EC_CURVES[curveOid]
                    ?: throw IllegalStateException("unsupported EC DSC curve $curveOid")
                val point = spki.publicKeyData.bytes // 0x04 || X || Y
                require(point.size == 1 + 2 * curve.coordLen && point[0].toInt() == 0x04) {
                    "expected an uncompressed ${curve.algorithm} DSC key, got ${point.size} bytes"
                }
                map["dsc_pubkey_x"] = byteArrayInput(point.copyOfRange(1, 1 + curve.coordLen))
                map["dsc_pubkey_y"] = byteArrayInput(point.copyOfRange(1 + curve.coordLen, 1 + 2 * curve.coordLen))
                // ECDSA sig DER SEQUENCE{r,s}; s normalized low-s (Noir/noir-ecdsa reject s>n/2).
                val (r, s) = decodeEcdsaDer(signer.signature)
                val sLow = if (s > curve.order.shiftRight(1)) curve.order.subtract(s) else s
                if (curve.algorithm == "lean_poa") { // P-256: std secp256r1 takes r‖s combined
                    map["sod_signature"] = byteArrayInput(toLen(r, curve.coordLen) + toLen(sLow, curve.coordLen))
                } else { // other curves: separate r / s
                    map["sod_signature_r"] = byteArrayInput(toLen(r, curve.coordLen))
                    map["sod_signature_s"] = byteArrayInput(toLen(sLow, curve.coordLen))
                }
                curve.algorithm
            }
            "1.2.840.113549.1.1.1" -> { // rsaEncryption
                val n = org.bouncycastle.asn1.pkcs.RSAPublicKey.getInstance(spki.parsePublicKey()).modulus
                val bits = when {
                    n.bitLength() in 2040..2048 -> 2048
                    n.bitLength() in 4088..4096 -> 4096
                    else -> throw IllegalStateException("unsupported RSA modulus size ${n.bitLength()}")
                }
                val limbs = bits / 120 + 1 // 18 for 2048, 35 for 4096 (noir-bignum 120-bit limbs)
                // Barrett reduction param, matching noir-bignum: floor(2^(2*bits+6)/n).
                val redc = BigInteger.ONE.shiftLeft(2 * bits + 6).divide(n)
                map["dsc_modulus"] = limbInput(n, limbs)
                map["dsc_redc"] = limbInput(redc, limbs)
                map["sod_signature"] = limbInput(BigInteger(1, signer.signature), limbs)
                "lean_poa_rsa$bits"
            }
            else -> throw IllegalStateException("unsupported DSC key algorithm ${spki.algorithm.algorithm.id}")
        }

        map["current_date"] = scalarInput(currentDateYymmdd)
        return Inputs(algorithm, map)
    }

    // ── helpers ──────────────────────────────────────────────────────────────

    private fun sha256(b: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(b)

    /** Big-endian magnitude bytes of a positive BigInteger (drops the sign byte). */
    private fun BigInteger.toByteArrayUnsigned(): ByteArray {
        val b = toByteArray()
        return if (b.isNotEmpty() && b[0].toInt() == 0) b.copyOfRange(1, b.size) else b
    }

    /** noir_android takes each [u8; N] as a list of per-byte hex strings. */
    private fun byteArrayInput(b: ByteArray): List<String> = b.map { "0x%02x".format(it.toInt() and 0xff) }

    /**
     * noir_android takes a scalar (u32/Field) as a hex string too, and enforces
     * it: Circuit.generateWitnessMap accepts a Number, or a String that starts
     * with "0x", and rejects anything else with "Expected hexadecimal number for
     * parameter: <name>". A decimal string is not a parse failure there — it is
     * a hard reject, so every scalar has to go through here.
     */
    private fun scalarInput(v: Int): String = "0x%x".format(v)

    private val LIMB_MASK = BigInteger.ONE.shiftLeft(120).subtract(BigInteger.ONE)

    /** Decomposes a BigInteger into `count` 120-bit little-endian limbs (noir-bignum layout). */
    private fun limbInput(x: BigInteger, count: Int): List<String> =
        (0 until count).map { "0x" + x.shiftRight(120 * it).and(LIMB_MASK).toString(16) }

    private fun pad(b: ByteArray, n: Int): ByteArray = if (b.size == n) b else b.copyOf(n)

    /** Left-pads a positive BigInteger to an `n`-byte big-endian array. */
    private fun toLen(v: BigInteger, n: Int): ByteArray {
        val raw = v.toByteArrayUnsigned()
        require(raw.size <= n) { "value exceeds $n bytes" }
        val out = ByteArray(n)
        System.arraycopy(raw, 0, out, n - raw.size, raw.size)
        return out
    }

    private fun decodeEcdsaDer(der: ByteArray): Pair<BigInteger, BigInteger> {
        val seq = ASN1Sequence.getInstance(der)
        val r = (seq.getObjectAt(0) as ASN1Integer).positiveValue
        val s = (seq.getObjectAt(1) as ASN1Integer).positiveValue
        return r to s
    }

    /** Strips the EF.SOD application tag (0x77) if present, yielding the ContentInfo. */
    private fun stripSodTag(sod: ByteArray): ByteArray {
        if (sod.isNotEmpty() && (sod[0].toInt() and 0xff) == 0x77) {
            var i = 1
            val lenByte = sod[i].toInt() and 0xff
            i += if (lenByte < 0x80) 1 else 1 + (lenByte and 0x7f)
            return sod.copyOfRange(i, sod.size)
        }
        return sod
    }

    private fun indexOfSubarray(haystack: ByteArray, needle: ByteArray): Int {
        if (needle.isEmpty() || haystack.size < needle.size) return -1
        outer@ for (i in 0..haystack.size - needle.size) {
            for (j in needle.indices) if (haystack[i + j] != needle[j]) continue@outer
            return i
        }
        return -1
    }
}

File: ios/EarthCore/Sources/EarthCore/Passport/PassportInputs.swift
import BigInt
import Foundation

/// Builds the `lean_poa` circuit inputs from a scanned passport's EF.DG1 and
/// EF.SOD. Ports `wallet/passport/PassportInputs.kt`.
///
/// Everything is derived from the passport itself. The circuit computes the
/// nullifier (Poseidon2 over name‖DOB) and the DSC commitment in-circuit and
/// returns them, so there is no Poseidon2 needed here and no registry round
/// trip before proving — the Document Signer certificate simply travels with
/// `MsgRegister` for the chain to check against its CSCA trust store.
///
/// Circuit input contract (see `circuits/lean_poa/SECURITY.md`):
///
///     dg1[95] + dg1_len
///     e_content[200] + e_content_len + dg1_hash_offset
///     signed_attrs[200] + signed_attrs_len + econtent_hash_offset
///     dsc_pubkey_x[32], dsc_pubkey_y[32], sod_signature[64] (low-s)
///     current_date, address
public enum PassportInputs {

    public enum Error: Swift.Error, Equatable {
        /// The account address did not decode to twenty bytes.
        case badAddress(Int)
        case dg1TooLong(Int)
        case eContentTooLong(Int)
        case signedAttributesTooLong(Int)
        case dg1HashNotInEContent
        case eContentHashNotInSignedAttributes
        case malformedSignature
    }

    static let dg1Max = 95
    static let eContentMax = 200
    static let signedAttributesMax = 200

    /// The circuit inputs, and which circuit they are for.
    public struct Inputs {
        /// `lean_poa`, `lean_poa_rsa2048`, … — selects both the compiled
        /// circuit to prove with and the chain's verifying key.
        public let algorithm: String
        /// The witness map, in the form the Noir binding takes: every value a
        /// hex string, or a list of per-byte hex strings.
        public let witness: [String: Any]
    }

    /// A Document Signer lifted out of a SOD: its DER, to travel with
    /// `MsgRegister`, and its canonical public key.
    public struct ScannedDSC {
        public let certificateDER: Data
        public let publicKey: Data
    }

    public static func scannedDSC(efSOD: Data) throws -> ScannedDSC {
        let sod = try SOD(efSOD: efSOD)
        return ScannedDSC(
            certificateDER: sod.certificate.der,
            publicKey: sod.certificate.canonicalPublicKey
        )
    }

    /// - Parameters:
    ///   - dg1: raw EF.DG1 as read from the chip.
    ///   - efSOD: raw EF.SOD.
    ///   - currentDateYYMMDD: today as a YYMMDD integer. The chain pins this to
    ///     block time within `current_date_max_skew_seconds`, so it must be
    ///     roughly now — see finding 7 in the circuit's SECURITY.md.
    ///   - address: the bech32 account this registration is for. The circuit
    ///     takes it as a public input, so the proof verifies only for this
    ///     account and cannot be lifted out of a block and replayed from another
    ///     wallet — which is what lets the chain treat a re-registration as
    ///     MOVING a registration rather than refusing it. It must be the account
    ///     that will sign MsgRegister; the chain rejects any other.
    public static func build(
        dg1: Data,
        efSOD: Data,
        currentDateYYMMDD: Int,
        address: String
    ) throws -> Inputs {
        guard dg1.count <= dg1Max else { throw Error.dg1TooLong(dg1.count) }

        let sod = try SOD(efSOD: efSOD)
        let eContent = sod.eContent
        let signedAttributes = sod.signedAttributes

        guard eContent.count <= eContentMax else { throw Error.eContentTooLong(eContent.count) }
        guard signedAttributes.count <= signedAttributesMax else {
            throw Error.signedAttributesTooLong(signedAttributes.count)
        }

        // The two hash bindings the circuit re-checks: sha256(dg1) sits inside
        // eContent, and sha256(eContent) sits inside the signed attributes as
        // the messageDigest attribute. The circuit is told where rather than
        // made to search, so the offsets are found here.
        guard let dg1HashOffset = index(of: Hashes.sha256(dg1), in: eContent) else {
            throw Error.dg1HashNotInEContent
        }
        guard let eContentHashOffset = index(of: Hashes.sha256(eContent), in: signedAttributes) else {
            throw Error.eContentHashNotInSignedAttributes
        }

        var witness: [String: Any] = [
            "dg1": byteArray(pad(dg1, to: dg1Max)),
            "dg1_len": scalar(dg1.count),
            "e_content": byteArray(pad(eContent, to: eContentMax)),
            "e_content_len": scalar(eContent.count),
            "dg1_hash_offset": scalar(dg1HashOffset),
            "signed_attrs": byteArray(pad(signedAttributes, to: signedAttributesMax)),
            "signed_attrs_len": scalar(signedAttributes.count),
            "econtent_hash_offset": scalar(eContentHashOffset),
            "current_date": scalar(currentDateYYMMDD),
            "address": try addressField(address),
        ]

        switch sod.certificate.publicKey {
        case let .ec(curve, x, y):
            witness["dsc_pubkey_x"] = byteArray(x)
            witness["dsc_pubkey_y"] = byteArray(y)

            let (r, s) = try decodeECDSASignature(sod.signature)
            // Noir's std ECDSA and noir-ecdsa both reject s > n/2 as malleable,
            // so s is normalised here. ICAO does not require the low form, and
            // roughly half of real signatures arrive in the high one.
            let lowS = s > curve.order >> 1 ? curve.order - s : s
            let width = curve.coordinateLength

            if curve.algorithm == "lean_poa" {
                // P-256 goes through Noir's std secp256r1, which takes r‖s as
                // one 64-byte input; the other curves take them apart.
                witness["sod_signature"] = byteArray(
                    leftPad(r, to: width) + leftPad(lowS, to: width)
                )
            } else {
                witness["sod_signature_r"] = byteArray(leftPad(r, to: width))
                witness["sod_signature_s"] = byteArray(leftPad(lowS, to: width))
            }

        case let .rsa(modulus, _):
            let bits = try rsaCircuitBits(modulus)
            // noir-bignum stores a big integer as 120-bit little-endian limbs.
            let limbs = bits / 120 + 1
            // Barrett reduction parameter, as noir-bignum defines it.
            let redc = (BigInt(1) << (2 * bits + 6)) / modulus
            let signature = BigInt(sign: .plus, magnitude: BigUInt(sod.signature))

            witness["dsc_modulus"] = limbArray(modulus, count: limbs)
            witness["dsc_redc"] = limbArray(redc, count: limbs)
            witness["sod_signature"] = limbArray(signature, count: limbs)
        }

        return Inputs(algorithm: try sod.certificate.registerAlgorithm, witness: witness)
    }

    // MARK: - helpers

    private static func rsaCircuitBits(_ modulus: BigInt) throws -> Int {
        switch modulus.magnitude.bitWidth {
        case 2040 ... 2048: return 2048
        case 4088 ... 4096: return 4096
        case let other: throw Certificate.Error.unsupportedModulusSize(other)
        }
    }

    /// The Noir binding takes every `[u8; N]` as a list of per-byte hex strings.
    static func byteArray(_ data: Data) -> [String] {
        data.map { String(format: "0x%02x", $0) }
    }

    /// Scalars are hex strings too, and the binding enforces it: a decimal
    /// string is not parsed loosely, it is rejected outright.
    static func scalar(_ value: Int) -> String { "0x" + String(value, radix: 16) }

    /// The twenty address bytes big-endian as one field element — the same
    /// encoding the chain compares against in `verifyRegistrationProof`.
    /// 160 bits into BN254's ~254-bit field, so the encoding is injective and
    /// two accounts can never collide onto one element.
    static func addressField(_ bech32: String) throws -> String {
        let (_, payload) = try Bech32.decode(bech32)
        guard payload.count == 20 else { throw Error.badAddress(payload.count) }
        return "0x" + payload.map { String(format: "%02x", $0) }.joined()
    }

    /// A big integer as `count` 120-bit little-endian limbs.
    static func limbArray(_ value: BigInt, count: Int) -> [String] {
        let mask = (BigInt(1) << 120) - 1
        return (0 ..< count).map { "0x" + String((value >> (120 * $0)) & mask, radix: 16) }
    }

    static func pad(_ data: Data, to length: Int) -> Data {
        data.count == length ? data : data + Data(repeating: 0, count: length - data.count)
    }

    /// Big-endian, left-padded to a fixed width. Dropping the padding would
    /// change the value for any coordinate with a leading zero byte.
    static func leftPad(_ value: BigInt, to length: Int) -> Data {
        let bytes = value.magnitude.serialize().drop { $0 == 0 }
        return Data(repeating: 0, count: length - bytes.count) + bytes
    }

    static func decodeECDSASignature(_ der: Data) throws -> (r: BigInt, s: BigInt) {
        guard let sequence = try? DER.parse(der).expect(tag: DER.Tag.sequence),
              let fields = try? sequence.children(), fields.count >= 2,
              fields[0].tag == DER.Tag.integer, fields[1].tag == DER.Tag.integer
        else { throw Error.malformedSignature }
        return (
            BigInt(sign: .plus, magnitude: BigUInt(fields[0].unsignedInteger)),
            BigInt(sign: .plus, magnitude: BigUInt(fields[1].unsignedInteger))
        )
    }

    static func index(of needle: Data, in haystack: Data) -> Int? {
        guard !needle.isEmpty, haystack.count >= needle.count else { return nil }
        let bytes = [UInt8](haystack)
        let target = [UInt8](needle)
        outer: for i in 0 ... (bytes.count - target.count) {
            for j in target.indices where bytes[i + j] != target[j] { continue outer }
            return i
        }
        return nil
    }
}
